### ISSUE TITLE
Fatal error in Gargoyle with a .z3 file

### ISSUE THREAD

**Issue by no1453** [garglk/garglk#284]
Not sure if this is something you care to track down, but I thought I'd submit it.  A z3 file for the Infocom game Stationfall ("Stationfall R107-870430.z3") will crash Gargoyle after two input moves: "take card" then "put card in slot".   The latest version of Windows Frotz does not crash, with the same z3 file and input.
![gargoyle prob](https://user-images.githubusercontent.com/3926817/29850582-f22a3196-8ceb-11e7-85de-14e554be98e0.png)
This is with the latest stable release of Gargoyle.

**Comment by cspiegel** [comment]
OK, this is actually a bug in Stationfall, but most interpreters are lax here and allow it (some issuing a warning if in strict mode, but still storing a value).

I'll update Bocfel to store a value instead of dying, and ensure it's updated here before the next release.

**Comment by no1453** [comment]
Thanks, nice bit of Infocom trivia to know, also.

**Comment by cspiegel** [comment]
In fact, I think this is an unknown bug, or at least I can't find any mention of it anywhere.  It also crashes (with a different property of 229) if you try to put the uniform in the slot.  If you try to put the watch in the slot, though, you get “The slot swallows the chronometer and then spits it back.”.

I dug through the disassembly, and it appears that the code first checks whether the item being inserted into the slot is a form or not; and if it's not a form, it determines whether to print out the “spit it back” message or, if it's too big, “Unsurprisingly, the *foo* doesn't fit the slot.”

The way it looks up an item's size, however, appears to be completely bogus.  It's looking up a property on the “Level Three” object (for some reason), but looking it up by the object number (the ID card is object 81, and the uniform is 229).  Properties above 64 are invalid, so this just can't work.  Most interpreters just overrun the property defaults table and grab some random data from memory, which will cause one or the other of the messages to be printed.  It seems purely cosmetic, since in either case the object isn't accepted by the slot.

I tried this with the Lost Treasures version of Stationfall, and both the watch and ID card got spit back, but inserting the uniform caused the game to hang.

In the C64 version, the uniform printed the “doesn't fit” message.  This is consistent with modern interpreters which overrun the bounds of the properties table.

Basically, given a random junk value, it has a 50% chance of spitting the object back and a 50% chance of being too large.  It seems to me coincidental that the uniform's result is too large, whereas the smaller objects get spit back, because the size of the ID card is -16128, the size of the uniform is 4101, and the size of the watch is 0.  There is no rhyme or reason to these values, which leads me to believe it's purely coincidence that the resulting output is reasonable.

Anyway, interesting bug to track down!

**Comment by kerrymg** [comment]
I took a quick look at the assembly code and I believe the problem is that the operands to the Get Object Property (GETP) instruction are reversed.

Instead of:

>`GETP #12,G67 >STACK    ; 0E3FE: 31 0C 77 00`

it should be

>`GETP G67,#12 >STACK    ; 0E3FE: 51 77 0C 00`*

The latter form reads property number 12 from the object in global variable 103 (0x67) which in the original bug report presumably contained object 81. Object 81 is the "ID card" and that object has property 12 set to 2 which seems to be some sort of size indicator.

[*] Note that swapping the operands changes the opcode byte.

**Comment by cspiegel** [comment]
That is without a doubt the issue.  I'm testing a local branch of Bocfel that includes support for patching games on the fly. Currently it only does Stationfall, but could be used for other games with bugs that can be fixed in the same way. It's limited because it can't change the size, so I'm not sure if it'll ever be useful beyond Stationfall... I'm still debating whether it's worth it.

**Comment by Nathan2** [comment]
Exactly the same bug is found in Beyond Zork (all 4 releases), when you try to put something in the jar of bubble liquid.

**Comment by cspiegel** [comment]
Fixed in #300.
